key.core/src/main/java/de/uka/ilkd/key/settings/ChoiceSettings.java

```diff
@@ -98,64 +94,46 @@ private static ImmutableSet<Choice> choiceMap2choiceSet(HashMap<String, String>
      * @param remove remove entries not present in <code>choiceNS</code>
      */
     public void updateChoices(Namespace<Choice> choiceNS, boolean remove) {
-        Iterator<Choice> it = choiceNS.allElements().iterator();
-        HashMap<String, Set<String>> c2C = new LinkedHashMap<String, Set<String>>();
-        Choice c;
-        Set<String> soc;
-        while (it.hasNext()) {
-            c = (Choice) it.next();
-            if (c2C.containsKey(c.category())) {
-                soc = c2C.get(c.category());
-                soc.add(c.name().toString());
-                c2C.put(c.category(), soc);
-            } else {
-                soc = new LinkedHashSet<String>();
-                soc.add(c.name().toString());
-                c2C.put(c.category(), soc);
-            }
+        // Translate the given namespace into a map of 'string -> list[string]'
+        HashMap<String, Set<String>> c2C = new LinkedHashMap<>();
+        for (Choice c : choiceNS.allElements()) {
+            Set<String> soc = c2C.computeIfAbsent(c.category(), k -> new LinkedHashSet<>());
+            soc.add(c.name().toString());
         }
+
+        // if there differences in the stored defaults, changed it accordingly
         if (!c2C.equals(category2Choices)) {
-            if (remove) {
-                category2Choices = c2C;
-                fireSettingsChanged();
+            var tmp = new HashMap<>(category2Choices);
+            if (!remove) {
+                tmp.putAll(c2C);
+                setChoiceCategories(tmp);
             } else {
-                category2Choices.putAll(c2C);
-                ProofSettings.DEFAULT_SETTINGS.saveSettings();
+                setChoiceCategories(c2C);
```

Comment:
Correct me if I'm wrong but either the comment or the code is wrong: `choiceNS` might contain more elements than `category2Choices`, so removing all from `category2Choices` if they are not contained in `choiceNS` is not equal to `choiceNS`.

E.g. `category2Choices = [a: 1, b: 2]`, `choiceNS = [a: 1, c: 3, d: 4]` should imo lead to `category2Choices = [a: 1]` and currently leads to `category2Choices = [a: 1, c: 3, d: 4]`

Reply:
I would expect the latter result you gave. And key `b` is only removed if the parameter is set to true. 

It updates the internal data structure by the found choices in the KeY files. 

BTW. This was refactored from this:

```
public void updateChoices(Namespace<Choice> choiceNS, boolean remove){
	Iterator<Choice> it = choiceNS.allElements().iterator();
	HashMap<String,Set<String>> c2C = new LinkedHashMap<String, Set<String>>();
	Choice c;
	Set<String> soc;
	while(it.hasNext()){
	    c=(Choice)it.next();
	    if(c2C.containsKey(c.category())){
		soc=c2C.get(c.category());
		soc.add(c.name().toString());
		c2C.put(c.category(),soc);
	    }else{
		soc = new LinkedHashSet<String>();
		soc.add(c.name().toString());
		c2C.put(c.category(),soc);
	    }
	}
	if(!c2C.equals(category2Choices)){
	    if(remove){
		category2Choices = c2C;
		fireSettingsChanged();
	    }else{
		category2Choices.putAll(c2C);
		ProofSettings.DEFAULT_SETTINGS.saveSettings();
	    }
	}
	for (final String s : getDefaultChoices().keySet()) {
	    if(category2Choices.containsKey(s)){
		if(!category2Choices.get(s).
		   contains(category2Default.get(s))){
		    category2Default.put(s,category2Choices.get(s).
					 iterator().next());
		    fireSettingsChanged();
		}
	    }else{
		category2Default.remove(s);
		fireSettingsChanged();
	    }
	}
    }

```